File: skillet/src/main/java/dev/kkorolyov/pancake/skillet/serialization/GenericEntitySerializer.java
package dev.kkorolyov.pancake.skillet.serialization;

import dev.kkorolyov.pancake.platform.serialization.string.StringSerializer;
import dev.kkorolyov.pancake.platform.serialization.string.entity.EntityStringSerializer;
import dev.kkorolyov.pancake.skillet.model.GenericComponent;
import dev.kkorolyov.pancake.skillet.model.GenericEntity;

import java.util.Arrays;
import java.util.regex.MatchResult;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

/**
 * Serializes {@link GenericEntity}.
 */
public class GenericEntitySerializer extends StringSerializer<GenericEntity> {
	private static final String PATTERN = new EntityStringSerializer(null).pattern();
	private static final String SPLIT_PATTERN = ",\\s*(?=\\w+\\s*\\{)";
	private static final Pattern NAME_PATTERN = Pattern.compile("\\w+(?=\\s*\\[)");
	private static final StringSerializer<GenericComponent> componentSerializer = new GenericComponentSerializer();

	/**
	 * Constructs a new generic entity serializer.
	 */
	public GenericEntitySerializer() {
		super(PATTERN);
	}

	@Override
	public GenericEntity read(String out) {
		String name = NAME_PATTERN.matcher(out).results()
				.findFirst()
				.map(MatchResult::group)
				.orElseThrow(() -> new IllegalArgumentException("Does not contain an entity name: " + out));
		Iterable<GenericComponent> components = Arrays.stream(out.split(SPLIT_PATTERN))	// Split beforehand because matches() is greedy
				.flatMap(componentSerializer::matches)
				.collect(Collectors.toList());

		return new GenericEntity(name, components);
	}

	@Override
	public String write(GenericEntity in) {
		return in.getName() + in.getComponents().stream()
				.map(componentSerializer::write)
				.collect(Collectors.joining("," + System.lineSeparator() + "\t", "[" + System.lineSeparator() + "\t", System.lineSeparator() + "]"));
	}
}

File: skillet/src/main/kotlin/dev/kkorolyov/pancake/skillet/serialization/GenericEntitySerializer.kt
package dev.kkorolyov.pancake.skillet.serialization

import dev.kkorolyov.pancake.platform.serialization.string.StringSerializer
import dev.kkorolyov.pancake.platform.serialization.string.entity.EntityStringSerializer
import dev.kkorolyov.pancake.skillet.model.GenericEntity
import java.util.stream.Stream

private val namePattern = """\w+""".toRegex()
private val baseBattern = EntityStringSerializer(null).pattern()
private val nameMatchPattern = """$namePattern(?=$baseBattern)""".toRegex()

/**
 * Serializes [GenericEntity].
 */
object GenericEntitySerializer : StringSerializer<GenericEntity>("$namePattern$baseBattern") {
	override fun read(out: String): GenericEntity =
			GenericEntity(
					nameMatchPattern.find(out)?.value ?: throw IllegalArgumentException("Does not contain an entity name: $out"),
					Iterable { GenericComponentSerializer.matches(out).iterator() }
			)
	override fun write(`in`: GenericEntity): String =
			`in`.name + `in`.components.joinToString(",${System.lineSeparator()}\t", "[${System.lineSeparator()}\t", "]") {
				GenericComponentSerializer.write(it)
			}

	override fun matches(out: String): Stream<GenericEntity> =
			out.split(""",\s*(?=${pattern()})""".toRegex()).stream()
					.flatMap { super.matches(it) }
}
